### DISCUSSION TITLE
Auto change day/night theme

### DISCUSSION THREAD

**Discussion by thedoctar** [lxqt/lxqt#2296] [Ideas]
Hi all,

I was wondering if we could add a feature to automatically change between day/night theme in lxqt. Currently it's not even possible to save qt palettes in the UI. I'm using Lubuntu 22.04.1.

To implement this, one could simply check the time on an app start-up and choose between two colour palettes. Then you don't have to run a process that always checks the time.

Best,

thedoctar

**Comment by stefonarch** [comment]
> Currently it's not even possible to save qt palettes in the UI. I'm using Lubuntu 22.04.1.

Still at 0.17? https://lubuntu.me/jammy-backports-22-04-1/

This would mean:
* define 2 icon sets (dark/bright)
* define theme eventually
* choose night/day palettes
* interaction with redshift or similar

**Reply by thedoctar** [reply]
Thanks! I enabled backports now.

I do this in vim using .vimrc to check the time and choose between the shine palette and slate.

Also it would be great to also do this for qterminal, as they also have pre-defined light/dark colour palettes.

We also have to auto-change the setting for apps that use gtk.

**Comment by tsujan** [comment]
Even ignoring the fact that on-the-fly changes to Qt style can cause complications in some apps and that not all Qt styles follow palettes, it would mean contradicting LXQt's philosophy.

**Reply by thedoctar** [reply]
Yes, but maybe this can be an optional addon? Currently you can still change the Qt pallette without restarting LXDE, no?

**Reply by tsujan** [reply]
> Yes, but maybe this can be an optional addon?

Optional or not, it'll have a unjustifiable cost.

> Currently you can still change the Qt pallette without restarting LXDE, no?

Please reread my previous comment.

**Reply by thedoctar** [reply]
If real-time switching already exists, it's a small step to set up an automatic process to change it. Light doesn't mean minimalistic. There are far lighter DEs than Lubuntu. E.g. antiX is way way lighter. It uses icewm.

**Reply by tsujan** [reply]
If it's just a small step, you might want to ask it from Lubuntu. LXQt has no relation to Lubuntu.

**Reply by thedoctar** [reply]
I did, but they told me to go upstream XD. If you don't think it's worth implementing upstream I will go back to them then.

**Reply by thedoctar** [reply]
Perhaps you can provide me with some advice. If I were to simply change lxqt.conf via a bash script and run the function StyleConfig::applyStyle(), I would apply the new settings in lxqt.conf. So all I need to do is run a cron job to check the time each hour say, and run a script to rewrite the settings and then apply the style. Would this break anything?

**Reply by tsujan** [reply]
For your personal use, yes, you could make changes to `lxqt.conf` by using a bash script. Change all palette values and, perhaps, `theme` and `icon_theme`. The changes to that file will be applied automatically. You just need to write a bash script that does it correctly.

You might see some kind of flickering with each change. If a Qt app shows weird behaviors after changes, just restart it (no need to restart the session). Some Qt/KDE apps (like Krita) might not change.

EDIT: Use a Qt style that responds to palette changes. Fusion (the default Qt style) and Breeze do.

**Reply by thedoctar** [reply]
Great! Many thanks!

**Reply by thedoctar** [reply]
Would this also work for the qterminal theme?

**Reply by tsujan** [reply]
No. It doesn't work with many things, but it should work for changing the colors of Fusion (or Breeze) in most apps. You could also change the wallpaper in your script.

**Comment by tsujan** [comment]
Small steps to implement daylight themes in your DE (expanding on what @stefonarch mentioned above):

 1. First of all, try to make your toolkit unstylable, as GNOME did to GTK (→ https://stopthemingmy.app/). If you can't control the toolkit, do what KDE devs do — they implicitly break `QStyle` in their codes and enforce Breeze plus their non-Qt color schemes (fortunately, they can't touch Qt directly).
 2. Then make your toolkit have two color variants. If you can't control the toolkit, make your DE dependent on them. Also, tell your users they shouldn't use this or that Qt style. You can also be smart and remain silent, until users encounter problems; then you could say to them, "What did you expect...?"
 3. Probably, enforce 2 icon sets and make them dependencies of your DE.
 4. Add codes to change the colors based on the time, not only by applying the color variants to all running apps but also by changing the themes of your panel, runner,… 
 5. If some Qt apps have problems with an on-the-fly theme change, tell your users to restart them or don't use them at all. Silence is preferable though.
 6. If your panel has a problem too, tell your users they should tolerate it because you can't do anything about Qt Stylesheets.
 7. Oh, don't forget this: you should make a window manager for your DE, so that you can change its colors too. If not, find one that has that property and enforce it (you can't have a WM-agnostic DE anymore).
 8. Don't forget to add a lot of alien codes for GTK.

**Reply by thedoctar** [reply]
I see your concern now. You're worried about other developers of apps using Qt or GTK complaining?

Okay, maybe one option is to set up a day/night cron job, which simply checks for day/night and runs a bunch of scripts. Each app developer could then write their own theme changer. There could be a central default theme-changing script for the Lubuntu/lxqt apps like qterminal, pcmanfm-qt and qpdfview. Simple apps could rely on that.

A very naïve definition of the night-time theme is the default theme with inverted colours.

**Reply by tsujan** [reply]
LXQt can't do it, for more than 8 reasons.

Lubuntu's decisions aren't related to LXQt, but if it adds such scripts, it should tell its users not to report problems to LXQt.

Any user with any distro can do it for himself/herself and take full responsibility.
